### ISSUE TITLE
[JAWS, Chrome] - alerts do not work in Fullscreen mode

### ISSUE THREAD

**Issue by loganbraywork** [phetsims/molarity#155]
Report a Problem
Instructions: Copy the form below and paste it into an email addressed to phethelp@colorado.edu. Please fill out the form as accurately and concisely as you can.

**Test device**
Windows 10 laptop
**Operating System**
Windows 10 JAWS
**Browser**
Chrome
**Problem description**
For https://github.com/phetsims/QA/issues/436

When in fullscreen mode, JAWS fails to read out full description of slider interactions. Will only read out the new numerical value of the slider but fails to describe anything else which occurs. Does not occur outside of fullscreen mode.
**Steps to reproduce**
1. Open JAWS on google chrome
2. Change sim to full screen
3. Using the arrow keys, manipulate the slider

Troubleshooting information:
!!!!! DO NOT EDIT !!!!!
Name: ‪Molarity‬
URL: https://phet-dev.colorado.edu/html/molarity/1.5.0-dev.23/phet/molarity_en_phet.html
Version: 1.5.0-dev.23 2019-10-04 22:27:51 UTC
Features missing: generatedcontent, touch
Flags: pixelRatioScaling
User Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.90 Safari/537.36
Language: en-US
Window: 1536x754
Pixel Ratio: 1.25/1
WebGL: WebGL 1.0 (OpenGL ES 2.0 Chromium)
GLSL: WebGL GLSL ES 1.0 (OpenGL ES GLSL ES 1.0 Chromium)
Vendor: WebKit (WebKit WebGL)
Vertex: attribs: 16 varying: 30 uniform: 4095
Texture: size: 16384 imageUnits: 16 (vertex: 16, combined: 32)
Max viewport: 32767x32767
OES_texture_float: true
Dependencies JSON: {}

**Comment by twant** [comment]
@terracoda @jessegreenberg @zepumph have any of you encountered this error with JAWS? I don't have JAWS installed on my machine currently.

**Comment by terracoda** [comment]
@twant, I am not familiar with this bug.

@loganbraywork did you first Tab to a slider and then use the Arrows keys to adjust one of the them?

**Comment by jessegreenberg** [comment]
I confirmed the issue. But it isn't specific to sliders, no alerts are working when in full screen mode.

**Comment by jessegreenberg** [comment]
This does not happen with NVDA in Firefox, but it does happen with NVDA in Chrome. This indicates it is a Chrome bug.

**Comment by jessegreenberg** [comment]
I made a quick test to learn more. In the following HTML, this bug does not happen:
<details>
<summary>Working Fullscreen Alerts</summary>
```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>title</title>
  </head>
  <body>
    <div id="parent">
      <p aria-live="polite" id='alert'></p>
      <button id='fullscreen-button'>Full Screen</button>
    </div>
  </body>

  <script type="text/javascript">
    const alert = document.getElementById( 'alert' );

    let counter = 1;
    window.setInterval( () => {
      alert.textContent = 'Alert: ' + counter++;
    }, 2000 );

    const button = document.getElementById( 'fullscreen-button' );
    button.addEventListener( 'click', () => {
      document.getElementById( 'parent' ).requestFullscreen();
    } );
  </script>
</html>
```
</details>

But in this HTML the bug does happen:
<details>
<summary>Broken fullscreen alerts</summary
```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>title</title>
  </head>
  <body>
    <p aria-live="polite" id='alert'></p>
    <div id="parent">
      <button id='fullscreen-button'>Full Screen</button>
    </div>
  </body>

  <script type="text/javascript">
    const alert = document.getElementById( 'alert' );

    let counter = 1;
    window.setInterval( () => {
      alert.textContent = 'Alert: ' + counter++;
    }, 2000 );

    const button = document.getElementById( 'fullscreen-button' );
    button.addEventListener( 'click', () => {
      document.getElementById( 'parent' ).requestFullscreen();
    } );
  </script>
</html>
```
</details>

So if the aria-live element is outside of the fullscreen element (which is the case in our sims), Chrome does not send the aria-live alert content to the screen reader.

The most direct fix for this would be to move the aria-live elements into the PDOM or at least under the Display domElement. I think this should work. This means moving the alert system more into scenery, so it would be good to think about how this may impact moving utteranceQueue to its own repository.

@terracoda how severe do you think this issue is and does it need to block publication of this sim and others?

**Comment by jessegreenberg** [comment]
Adding a11y-dev meeting label to review proposed fix or consider others.

**Comment by zepumph** [comment]
I think the best way forward is to move UtteranceQueue to its own repo, and move aria live tags into scenery's display.

To me it makes sense that Interactive descriptions as a whole are controlled by scenery, as they together create this feature. I worry that, after encountering this, if we do nothing, there will be more cases that come up in which we want to more closely control aria-live and the PDOM together as one system. 

Furthermore, you could make a substantial argument that scenery, as a library, would definitely benefit from this change.

**Comment by zepumph** [comment]
>@terracoda how severe do you think this issue is and does it need to block publication of this sim and others?

I also think it is important for @emily-phet to weight in on this, as it would be a fair bit of time to move over UtteranceQueue and make sure proper initialization was happening in scenery/Display.js

**Comment by emily-phet** [comment]
@terracoda I'd love to hear your thoughts on this before I join in.

**Comment by jessegreenberg** [comment]
Discussed with @zepumph this afternoon - we definitely think that moving utteranceQueue to its own repo and integrating with scenery is worthwhile and the right thing to do. @zepumph also talked with @terracoda and @terracoda was on board with this, but thought we should be cautious about regressions.

We recommend fixing this in time for the molarity publication, but not patching into the gravity-force-lab-basics publication. @zepumph and @twant will work on this together soon, but if @terracoda or @emily-phet have concerns please let us know.

**Comment by terracoda** [comment]
Yes, @zepumph and I discussed this.
@jessegreenberg and I have had conversations before about where the aria-live content should go.
In my original design thinking, alert content (mainly context responses) were not ever to be accessible in the PDOM because the alert content can get stale quickly.

@jessegreenberg has done some things to ensure that the stale alerts are not accidentally accessed upon a re-read of the sim's PDOM.

I do think that the having aria-live alerts outside of the PDOM makes sense from an interaction design point of view.

**Comment by terracoda** [comment]
And I was able to reproduce the error on MacOs with Safari and Voice Over. It's a serious problem to have no alerts. Moving the aria-live content to a place where they will work in full screen mode, makes sense to me.

**Comment by zepumph** [comment]
https://github.com/phetsims/utterance-queue/issues/1 should have fixed this. Marking ready for review.

**Comment by KatieWoe** [comment]
This seems to be fixed in https://github.com/phetsims/QA/issues/455

**Comment by zepumph** [comment]
Hooray! Thanks.

**Comment by twant** [comment]
Thanks @KatieWoe! Closing this issue.
